File: src/classes/okx.ts
import { IBot } from "@/models/bot";
import { ensureDirExists } from "@/utils/orders/funcs";
import { getInterval, parseDate, parseFilledOrder, getExactDate } from "@/utils/funcs2";
import { botLog, getSymbol, timedLog } from "@/utils/functions";
import { writeFileSync } from "fs";
import { RestClient, WebsocketClient } from "okx-api";
import type { AlgoOrderResult, OrderDetails, OrderResult } from "okx-api";
import { DEV } from "@/utils/constants";
import { configDotenv } from "dotenv";
import { IOrderDetails } from "@/utils/interfaces";
configDotenv();

export class OKX {
    bot: IBot;
    flag: "1" | "0";
    apiKey: string;
    apiSecret: string;
    passphrase: string;

    client: RestClient;
    ws: WebsocketClient | null = null;

    constructor(bot: IBot) {
        console.log(
            `\nInit OKX For Bot=${bot.name}\tMode=${
                bot.demo ? "demo" : "live"
            }\n`
        );
        this.bot = bot;
        this.flag = this.bot.demo ? "1" : "0";
        this.apiKey = this.bot.demo
            ? process.env.OKX_API_KEY_DEV!
            : process.env.OKX_API_KEY!;
        this.apiSecret = this.bot.demo
            ? process.env.OKX_API_SECRET_DEV!
            : process.env.OKX_API_SECRET!;
        this.passphrase = process.env.OKX_PASSPHRASE!;

        this.client = new RestClient(
            {
                apiKey: this.apiKey,
                apiSecret: this.apiSecret,
                apiPass: this.passphrase,
            },
            this.bot.demo ? "demo" : "prod"
        );

        botLog(this.bot, "OKX INITIALIZED");
    }

    async getBal(ccy?: string) {
        console.log(`\nGETTING BALANCE FOR BOT=${this.bot.name}\n`);
        try {
            const res = await this.client.getBalance(ccy ?? this.bot.ccy);
            return Number(res[0].details[0].availBal);
        } catch (error) {
            console.log(error);
        }
    }

    async getTicker() {
        botLog(this.bot, "GETTING TICKER...");
        const res = await this.client.getTicker(this.getSymbol());
        const ticker = Number(res[0].last);
        console.log({ ticker });
        return ticker;
    }

    async cancelOrder({ ordId, isAlgo }: { ordId: string; isAlgo?: boolean }) {
        try {
            botLog(this.bot, "CANCELLING ORDER...");
            const res = await (isAlgo
                ? this.client.cancelAlgoOrder([
                      { algoId: ordId, instId: this.getSymbol() },
                  ])
                : this.client.cancelOrder({
                      ordId,
                      instId: this.getSymbol(),
                  }));

            if (res[0].sCode != "0") {
                botLog(this.bot, "FAILED TO CANCEL ORDER");
                console.log(res[0]);
                return;
            }
            return ordId;
        } catch (error) {}
    }
    async placeOrder(
        amt: number,
        price?: number,
        side: "buy" | "sell" = "buy",
        sl?: number,
        clOrderId?: string
    ) {
        /* Place limit order at previous close */
        const od = { price, sl, amt, side };
        botLog(this.bot, `PLACING ORDER: ${JSON.stringify(od)}`);
        try {
            let res: OrderResult[] | AlgoOrderResult[];

            if (side == "buy") {
                res = await this.client.submitOrder({
                    instId: this.getSymbol(),
                    tdMode: "cash",
                    ordType: "market",
                    side,
                    sz: amt.toString(),
                    clOrdId: clOrderId,
                    //px: price.toString(),
                });
            } else {
                if (price) {
                    res = await this.client.placeAlgoOrder({
                        instId: this.getSymbol(),
                        tdMode: "cash",
                        ordType: "conditional",
                        tpTriggerPx: price.toString(),
                        //slTriggerPx: sl.toString(),
                        side,
                        sz: amt.toString(),
                        tpOrdPx:
                            this.bot.order_type == "Market"
                                ? "-1"
                                : (price * (1 - 0.0 / 100)).toString(),
                        /* slOrdPx:
                              this.bot.order_type == "Market"
                                  ? "-1"
                                  : (sl * (1 - 0.0 / 100)).toString(), */
                        algoClOrdId: clOrderId,
                    });
                } else {
                    res = await this.client.submitOrder({
                        instId: this.getSymbol(),
                        tdMode: "cash",
                        ordType: "market",
                        side,
                        sz: amt.toString(),
                        clOrdId: clOrderId,
                        //px: price.toString(),
                    });
                }
            }

            if (res[0].sCode != "0") {
                console.log(res[0]);
                return;
            }
            console.log(`\ORDER PLACED FOR BOT=${this.bot.name}\n`);
            const d: any = res[0];
            const id: string =
                side == "buy" ? d.ordId : price ? d.algoId : d.ordId;
            return id;
        } catch (error) {
            console.log(error);
        }
    }

    async getOrderbyId(
        orderId: string,
        isAlgo = false,
        pair?: string[]
    ): Promise<IOrderDetails | null | "live" | undefined> {
        try {
            pair = pair ?? [this.bot.base, this.bot.ccy];
            let data: IOrderDetails | null = null;
            let finalRes: OrderDetails | null = null;
            console.log(this.bot.name, `IS_ALGO: ${isAlgo}`, orderId);

            const symbo = getSymbol(pair, this.bot.platform);
            const res = isAlgo
                ? await this.client.getAlgoOrderDetails({ algoId: orderId })
                : await this.client.getOrderDetails({
                      ordId: orderId!,
                      instId: symbo,
                  });
            if (DEV) {
                console.log(`DEV: ${this.bot.name}`);
                console.log(res);
            }
            if (isAlgo && res[0].state == "effective") {
                botLog(this.bot, "IS_EFFECTIVE");
                return await this.getOrderbyId(res[0].ordId);
            } else if (!isAlgo) {
                if (res[0].state == "live") return "live";
                else if (res[0].state == "filled") finalRes = res[0];
            }
            if (!finalRes) {
                botLog(this.bot, "[OKX Class] ORDER NOT YET FILLED");
                return "live";
            }
            //console.log(this.bot.name, "FINAL RES", finalRes);
            data = parseFilledOrder(finalRes, this.bot.platform);

            return data;
        } catch (error: any) {
            botLog(this.bot, "ERROR");
            botLog(this.bot, error);
            if (isAlgo && error?.code == "51603")
                return await this.getOrderbyId(orderId);
        }
    }

    async setLev(val: number) {
        const res = await this.client.setLeverage({
            instId: this.getSymbol(),
            mgnMode: "isolated",
            lever: `${val}`,
        });
        console.log(res);
    }

    async getKline() {
        const end = Date.now();
        return await this.getKlines({ end, limit: 1 });
    }
    async getKlines({
        start,
        end,
        interval,
        limit = 100,
        pair,
    }: {
        end?: number;
        start?: number;
        interval?: number;
        pair?: string[];
        limit?: number;
    }) {
        
        
        let klines: any[] = [];
        let cnt = 0;
        interval = interval ?? this.bot.interval;
        end = end ?? getExactDate(interval).getTime() - interval * 60 * 1000; 
        end += (interval * 60000)
        const symbol = pair
            ? getSymbol(pair, this.bot.platform)
            : this.getSymbol();
        console.log("GETTING KLINES FOR:", symbol);
        botLog(this.bot, "GETTING KLINES.. FOR " + symbol);

        const res = await this.client.getCandles(
            symbol,
            getInterval(interval, "okx"),
            {
                before: start ? `${start}` : undefined,
                after: end ? `${end}` : undefined,
            }
        );
        let data = res;
        klines = [...data].reverse();

        let d = [...klines];

        const last = Number(d[d.length - 1][0]);

        botLog(this.bot, { end: parseDate(end), last: parseDate(last) });
        if (end >= last + 2 * interval * 60000) {
            botLog(this.bot, "END > LAST");
            end -= interval * 60000
            return await this.getKlines({ start, end, interval, pair, limit });
        }
        const lastCandle = d[d.length - 1];

        return limit == 1 ? d[d.length - 1] : d;
    }

    getSymbol() {
        return `${this.bot.base}-${this.bot.ccy}`;
    }

    async getCurrencies() {
        try {
            const res = await this.client.getCurrencies();
            return res;
        } catch (e) {
            console.log(e);
        }
    }
}

File: classes/OKX.py
from typing import Literal, Optional
from dotenv import load_dotenv
from okx import MarketData, Account, Trade
from models.bot_model import Bot
from utils.constants import *
import json, os
from datetime import datetime

from utils.types import OrderData

load_dotenv()
klines = []


class OKX:

    ws_url = "wss://wspap.okx.com:8443/ws/v5/business?brokerId=9999"
    ws_url_private = "wss://wspap.okx.com:8443/ws/v5/private?brokerId=9999"
    inst = None

    def __init__(self, bot: Bot) -> None:
        
        print("INITIALIZE OKX")
        print(os.getenv("OKX_PASSPHRASE"))
        self.bot = bot
        self.flag = (
            "1" if self.bot.demo else "0"
        )  # Production trading:0 , demo trading:1
        self.api_key = os.getenv("OKX_API_KEY_DEV" if self.bot.demo else "OKX_API_KEY")
        self.api_secret = os.getenv(
            "OKX_API_SECRET_DEV" if self.bot.demo else "OKX_API_SECRET"
        )
        self.passphrase = os.getenv("OKX_PASSPHRASE")
        api_key, api_secret, passphrase = self.api_key, self.api_secret, self.passphrase

        self.passphrase = os.getenv("OKX_PASSPHRASE")
        self.acc_api = Account.AccountAPI(
            api_key, api_secret, passphrase, False, self.flag
        )
        self.trade_api = Trade.TradeAPI(
            api_key, api_secret, passphrase, False, self.flag
        )
        self.market_data_api = MarketData.MarketAPI(flag=self.flag)

    def get_klines(self, end=None):
        """Returns Reversed klines"""
        from utils.functions import get_interval

        klines = []
        print("GETTING OKX KLINES...")
        end = end if end else round(datetime.now().timestamp() * 1000)
        interval = self.bot.interval
        p = self.bot.platform
        res = self.market_data_api.get_candlesticks(
            instId=self.get_symbol(), bar=get_interval(interval, "okx"), after=str(int(end))
        )
        data = res["data"]
        klines = [*klines, *data]
        d = klines.copy()
        d.reverse()

        return d

    def get_bal(self, ccy):
        from utils.functions import err_handler

        try:
            print("GETTING BALANCE...")
            res = self.acc_api.get_account_balance(ccy=ccy)
            bal = res["data"][0]["details"][0]["availBal"]
            return float(bal)

        except Exception as e:
            print("FAILED TO GET BALANCE")
            err_handler(e)

    def get_symbol(self):
        app = self.bot
        return f"{app.base}-{app.ccy}"

    def place_order(
        self,
        amt: float,
        sl: float,
        price: float,
        cl_order_id: str,
        side: Literal["buy", "sell"] = "buy",
    ):
        try:
            from utils.functions import err_handler, bot_log
            from utils.funcs3 import fix_str
            
            od = {"price": price, "sl": sl, "amt": amt, "side": side}
            bot_log(self.bot, f"PLACING ORDER: {od}")
            cl_order_id = fix_str(cl_order_id)
            if side == "buy":
                res = self.trade_api.place_order(
                    instId=self.get_symbol(),
                    tdMode="cash",
                    ordType=self.bot.order_type.lower(),
                    side=side,
                    sz=amt,
                    px=str(price),
                    clOrdId=cl_order_id,
                )

            else:
                res = self.trade_api.place_algo_order(
                    instId=self.get_symbol(),
                    tdMode="cash",
                    side=side,
                    ordType="oco",
                    sz=amt,
                    tpTriggerPx=str(price),
                    tpOrdPx="-1",
                    slTriggerPx=str(sl),
                    slOrdPx="-1",
                    algoClOrdId=cl_order_id,
                )

            if res["code"] != "0":
                print(res)
                raise Exception("Failed to place order")

            print(f"{'OCO' if side == 'sell' else ''} ORDER PLACED SUCCESSFULLY!\n")
            data = res["data"][0]

            order_id = data["ordId"] if side == "buy" else data["algoId"]
            return order_id

        except Exception as e:
            err_handler(e)

    def get_order_by_id(self, order_id=None, is_algo=False):
        from utils.functions import err_handler, bot_log

        try:
            print(f"GETTING ORDER {order_id}")

            data: Optional[OrderData] = None
            res = (
                self.trade_api.get_algo_order_details(algoId=order_id)
                if is_algo
                else self.trade_api.get_order(instId=self.get_symbol(), ordId=order_id)
            )
            status = res["code"]
            if status == "0":
                d = res["data"][0]

                if is_algo and d["state"] == "effective":
                    return self.get_order_by_id(d["ordId"], is_algo=False)

                if not is_algo and d["state"] == "filled":
                    data = {
                        "id": d["ordId"],
                        "fill_px": float(d["avgPx"]),
                        "fill_sz": float(d["accFillSz"]),
                        "fee": float(d["fee"]),
                        "fill_time": float(d["fillTime"]),
                    }

                else:
                    bot_log(self.bot, "[OKX CLASS] ORDER NOT YET FILLED")
                    return "live"
                
                return data

        except Exception as e:
            err_handler(e)


# okx : OKX = None
